## 이미지와 레이블이 동일한 폴더 동일한 파일이 존재하는지 확인

In [1]:
import os

trainPath = '/home/jupyter/highway/val/'

trainImagePath = os.path.join(trainPath,'images') 
trainLabelPath = os.path.join(trainPath,'labels') 

In [2]:
trainImagePath, trainLabelPath

('/home/jupyter/highway/val/images', '/home/jupyter/highway/val/labels')

## 이미지의 폴더와 레이블의 폴더의 이름과 갯수가 동일한지
- 작성자는 폴더에서 이미 데이터를 꺼내놔서 false_count로 올라감

In [3]:
ImagefileName = os.listdir(trainImagePath)
LabelfileName = os.listdir(trainLabelPath)

In [5]:
true_count=0
false_count=0
for fileName in ImagefileName:
    if fileName in LabelfileName:
        true_count+=1
    else:
        false_count+=1

print(true_count)
print(false_count)

0
3333


In [6]:
fileName = ImagefileName[0]
fileName

'Suwon_CH04_20200722_1600_WED_9m_NH_highway_OW5_rainy_FHD_162.png'

In [7]:
imagefullPath = os.path.join('/home/jupyter/highway/val/images/'+ fileName)
imagefullPath

'/home/jupyter/highway/val/images/Suwon_CH04_20200722_1600_WED_9m_NH_highway_OW5_rainy_FHD_162.png'

## 파일명이 동일한지 확인 (폴더 한개 확인) 
-  작성자는 images 폴더안에 데이터를 모두 풀어넣어서 해당 코드로 한번에 확인함

In [50]:
from glob import glob

fileName = ImagefileName[0]
imagefullPath = os.path.join('/home/jupyter/highway/val/', 'images')
labelfullPath = os.path.join('/home/jupyter/highway/val/', 'labels')

#print(fullPath+'/*.png')
imagefiles_png = glob(imagefullPath+'/*.png')
imagefiles_jpg = glob(imagefullPath+'/*.jpg')

labelfiles = glob(labelfullPath+'/*.txt')
# imagefiles
# labelfiles
# imagefiles에서 확장자 제거 + 앞에 경로 제거
split_imagefiles =[]

# png 파일의 확장자를 제거하고 이름만 리스트에 담기
for imagefile in imagefiles_png:
    split_imagefiles.append(os.path.basename(os.path.splitext(imagefile)[0]))
    
# jpg 파일의 확장자를 제거하고 이름만 리스트에 담기
for imagefile in imagefiles_jpg:
    split_imagefiles.append(os.path.basename(os.path.splitext(imagefile)[0]))


print('iamge 총 개수:',len(split_imagefiles))
# split_imagefiles

iamge 총 개수: 3333


## val 데이터를 전처리하다가 labels 폴더 하위의 항목이 1개 더많은것 확인
- images 폴더 하위의 txt 파일 개수 : 3333
- labes 폴더 하위의 txt 파일 개수 : 3334

In [54]:
os.path.splitext(labelfile)[1]

'.txt'

### 차집합 계산식을 활용해 결측데이터 검출
- 점프 투 파이썬/02-6. 집합 자료형/교집합, 합집합, 차집합 구하기 참조

In [46]:
s1 = set(split_imagefiles)
len(s1)
s2 = set(split_labelfiles)
len(s2)
s2-s1

{'Suwon_CH02_20200721_2030_TUE_9m_NH_highway_TW5_sunny_FHD_001'}

### 검출 결과 labels 파일의 아래 파일이 images 파일에 없는것 확인
- 'Suwon_CH02_20200721_2030_TUE_9m_NH_highway_TW5_sunny_FHD_001'
### 해당 txt 파일 삭제

In [47]:
os.remove('/home/jupyter/highway/val/labels/Suwon_CH02_20200721_2030_TUE_9m_NH_highway_TW5_sunny_FHD_001.txt')

In [51]:
# labelfiles에서 확장자 제거 + 앞에 경로 제거
split_labelfiles = []

test_ext_cnt = 0
for labelfile in labelfiles:
    split_labelfiles.append(os.path.basename(os.path.splitext(labelfile)[0]))   
print('label 총 개수:',len(split_labelfiles))
#split_labelfiles

label 총 개수: 3333


In [52]:
true_count=0
false_count=0

false_list = []

for imagefile in split_imagefiles:
    if imagefile in split_labelfiles:
        true_count +=1
    else:
        false_count +=1
        false_list.append(imagefile)

print(true_count)
print(false_count)

if true_count == len(split_imagefiles):
    print("파일명이 전부 일치합니다!! 축하합니다!")

# print(false_list)

3333
0
파일명이 전부 일치합니다!! 축하합니다!


## 폴더 내부의 파일명이 동일한지 확인 (모든 폴더 확인)

### 작성자는 images 폴더 안에 폴더/.png 파일 형식이 아니라 images 폴더 안에 png 파일을 전부 옮겨는 작업을 수행했기 때문에 아래 작업은 생략함

In [12]:
from tqdm import tqdm

In [54]:
from glob import glob

for folderName in ImagefolderName:
    imagefullPath = os.path.join('/home/jupyter/myData/train/images/'+ folderName)
    labelfullPath = os.path.join('/home/jupyter/myData/train/labels/'+ folderName)
    #print(fullPath+'/*.png')
    imagefiles = glob(imagefullPath+'/*.png')
    labelfiles = glob(labelfullPath+'/*.txt')
    imagefiles
    # imagefiles에서 확장자 제거 + 앞에 경로 제거
    split_imagefiles =[]
    for imagefile in imagefiles:
        split_imagefiles.append(os.path.basename(os.path.splitext(imagefile)[0]))
    #split_imagefiles

    # labelfiles에서 확장자 제거 + 앞에 경로 제거
    split_labelfiles =[]
    for labelfile in labelfiles:
        split_labelfiles.append(os.path.basename(os.path.splitext(labelfile)[0]))
    #split_labelfiles

    true_count=0
    false_count=0
    for imagefile in split_imagefiles:
        if imagefile in split_labelfiles:
            true_count +=1
        else:
            false_count +=1
    if false_count>0 or true_count==0:
        print("===================")
        print(folderName)
        print(true_count)
        print(false_count)
        print("===================")

  0%|          | 3/24101 [00:17<39:33:03,  5.91s/it]


KeyboardInterrupt: 

## txt파일명이 잘못된 것 수정
### txt파일명에 언더바 대신 공백이 있음
### 이미지 이름은 'Suwon_CH02_20200721_2130_TUE_9m_NH_highway_TW5_sunny_FHD'
Suwon_CH02_20200721_2130_TUE_9m_NH_highway_TW5_sunny_FHD 001.txt

In [38]:
# falseLabels = glob('/home/jupyter/highway/train/labels/Suwon_CH02_20200721_2130*.txt')
# falseLabels

In [37]:
falseLabels = glob('/home/jupyter/highway/train/labels/Suwon_CH02_20200721_2130*.txt')

for falseLabel in falseLabels:
    newLabel = falseLabel.replace(' ','_')
    print(newLabel)
    os.rename(falseLabel, newLabel)

/home/jupyter/highway/train/labels/Suwon_CH02_20200721_2130_TUE_9m_NH_highway_TW5_sunny_FHD_068.txt
/home/jupyter/highway/train/labels/Suwon_CH02_20200721_2130_TUE_9m_NH_highway_TW5_sunny_FHD_132.txt
/home/jupyter/highway/train/labels/Suwon_CH02_20200721_2130_TUE_9m_NH_highway_TW5_sunny_FHD_113.txt
/home/jupyter/highway/train/labels/Suwon_CH02_20200721_2130_TUE_9m_NH_highway_TW5_sunny_FHD_040.txt
/home/jupyter/highway/train/labels/Suwon_CH02_20200721_2130_TUE_9m_NH_highway_TW5_sunny_FHD_060.txt
/home/jupyter/highway/train/labels/Suwon_CH02_20200721_2130_TUE_9m_NH_highway_TW5_sunny_FHD_144.txt
/home/jupyter/highway/train/labels/Suwon_CH02_20200721_2130_TUE_9m_NH_highway_TW5_sunny_FHD_016.txt
/home/jupyter/highway/train/labels/Suwon_CH02_20200721_2130_TUE_9m_NH_highway_TW5_sunny_FHD_045.txt
/home/jupyter/highway/train/labels/Suwon_CH02_20200721_2130_TUE_9m_NH_highway_TW5_sunny_FHD_030.txt
/home/jupyter/highway/train/labels/Suwon_CH02_20200721_2130_TUE_9m_NH_highway_TW5_sunny_FHD_139.txt


### 파일명 수정 후에도 train, rabel 데이터 개수 불일치 확인
### labels 폴더에 txt파일이 3개 더 많은것을 확인한 후 결측치인지 확인하기

In [57]:
print('iamge 총 개수:',len(split_imagefiles))
# split_imagefiles

print('label 총 개수:',len(split_labelfiles))
#split_labelfiles

iamge 총 개수: 24101
label 총 개수: 24104


In [64]:
imagesPath_test = glob('/home/jupyter/highway/train/images/*')
labelsPath_test = glob('/home/jupyter/highway/train/labels/*')
len(imagesPath_test), len(labelsPath_test)
#imagesPath_test

(24101, 24104)

In [72]:
iist_images[0]

'Suwon_CH03_20201213_0835_SUN_9m_NH_highway_OW5_snow_FHD_052.png'

In [74]:
from glob import glob

fileName = ImagefileName[0]
imagefullPath = os.path.join('/home/jupyter/highway/train/', 'images')
labelfullPath = os.path.join('/home/jupyter/highway/train/', 'labels')

#print(fullPath+'/*.png')
imagefiles = glob(imagefullPath+'/*')
labelfiles = glob(labelfullPath+'/*')

# imagefiles
# labelfiles
# imagefiles에서 확장자 제거 + 앞에 경로 제거
split_imagefiles = []
split_labelfiles = []

# images 파일의 확장자를 제거하고 이름만 리스트에 담기
for imagefile in imagefiles:
    split_imagefiles.append(os.path.basename(os.path.splitext(imagefile)[0]))
    
# labels 하위 파일의 확장자를 제거하고 이름만 리스트에 담기
for labelfile in labelfiles:
    split_labelfiles.append(os.path.basename(os.path.splitext(labelfile)[0]))

print(len(split_imagefiles), len(split_labelfiles))

24101 24104


### labels 폴더에서 아래 3개의 파일이 images 파일에 없는것을 확인
- Suwon_CH02_20200721_1530_TUE_9m_NH_highway_TW5_sunny_FHD_069.png
- uwon_CH03_20200722_1900_WED_9m_NH_highway_OW5_sunny_FHD_051.png
- Suwon_CH03_20200722_1900_WED_9m_NH_highway_OW5_sunny_FHD_052.png
### 원본 압축파일에서 확인해본 결과 해당 파일이 없는것을 확인

### 불일치 파일 삭제하기

In [79]:
true_count=0
false_count=0
for label in split_labelfiles:
    if label in split_imagefiles:
        true_count+=1
    else:
        false_count+=1
        print('불일치 파일: {}'.format(label))
        falsefilePath = os.path.join(labelfullPath,label + '.txt')
        #print(falsefilePath)
        #해당파일 삭제
        os.remove(falsefilePath)
        print('파일 삭제 완료!!!')
        

print(f'true_count:{true_count}')
print(f'false_count:{false_count}')

불일치 파일: Suwon_CH02_20200721_1530_TUE_9m_NH_highway_TW5_sunny_FHD_069
파일 삭제 완료!!!
불일치 파일: Suwon_CH03_20200722_1900_WED_9m_NH_highway_OW5_sunny_FHD_051
파일 삭제 완료!!!
불일치 파일: Suwon_CH03_20200722_1900_WED_9m_NH_highway_OW5_sunny_FHD_052
파일 삭제 완료!!!
true_count:24101
false_count:3


### 다시 파일 개수 확인하기

In [80]:
from glob import glob

fileName = ImagefileName[0]
imagefullPath = os.path.join('/home/jupyter/highway/train/', 'images')
labelfullPath = os.path.join('/home/jupyter/highway/train/', 'labels')

#print(fullPath+'/*.png')
imagefiles = glob(imagefullPath+'/*')
labelfiles = glob(labelfullPath+'/*')

# imagefiles
# labelfiles
# imagefiles에서 확장자 제거 + 앞에 경로 제거
split_imagefiles = []
split_labelfiles = []

# images 파일의 확장자를 제거하고 이름만 리스트에 담기
for imagefile in imagefiles:
    split_imagefiles.append(os.path.basename(os.path.splitext(imagefile)[0]))
    
# labels 하위 파일의 확장자를 제거하고 이름만 리스트에 담기
for labelfile in labelfiles:
    split_labelfiles.append(os.path.basename(os.path.splitext(labelfile)[0]))

print(len(split_imagefiles), len(split_labelfiles))

24101 24101


In [81]:
true_count=0
false_count=0
for label in split_labelfiles:
    if label in split_imagefiles:
        true_count+=1
    else:
        false_count+=1
        print('불일치 파일: {}'.format(label))
        falsefilePath = os.path.join(labelfullPath,label + '.txt')
        #print(falsefilePath)
        #해당파일 삭제
        os.remove(falsefilePath)
        print('파일 삭제 완료!!!')
        

print(f'true_count:{true_count}')
print(f'false_count:{false_count}')

true_count:24101
false_count:0
